File: crates/kuben-platform/src/duration.rs
//! Human durations used in CRD fields: `90s`, `15m`, `168h`, `14d`, `1h30m`.

use std::time::Duration;

/// Parse a duration made of `<integer><unit>` segments (`s`, `m`, `h`, `d`).
/// Returns `None` for empty, malformed or overflowing input.
#[must_use]
pub fn parse(input: &str) -> Option<Duration> {
    let s = input.trim();
    if s.is_empty() {
        return None;
    }
    let mut total: u64 = 0;
    let mut digits = String::new();
    for c in s.chars() {
        if c.is_ascii_digit() {
            digits.push(c);
            continue;
        }

File: internal/kube/controller/duration.go
package controller

import (
	"math"
	"strconv"
	"strings"
)

// parseDuration reads the human durations of CRD fields (crate::duration):
// `<integer><unit>` segments with the units s, m, h and d, e.g. `90s`,
// `15m`, `168h`, `14d`, `1h30m`. It returns whole seconds, and false for
// empty, malformed or overflowing input.
func parseDuration(input string) (uint64, bool) {
	s := strings.TrimSpace(input)
	if s == "" {
		return 0, false
	}
	var total uint64
	start := 0
	for i := range len(s) {
		c := s[i]
		if c >= '0' && c <= '9' {
			continue
		}
		var unit uint64
		switch c {
		case 's':
			unit = 1
		case 'm':
			unit = 60
		case 'h':
			unit = 3_600
		case 'd':
			unit = 86_400
		default:
			return 0, false
		}
		n, err := strconv.ParseUint(s[start:i], 10, 64)
		if err != nil {
			return 0, false
		}
		start = i + 1
		if n > math.MaxUint64/unit || total > math.MaxUint64-n*unit {
			return 0, false
		}
		total += n * unit
	}
	if start != len(s) {
		return 0, false // trailing number without unit
	}
	return total, true
}

// secondsToMillis is seconds in milliseconds, saturating at MaxInt64.
func secondsToMillis(seconds uint64) int64 {
	if seconds > math.MaxInt64/1000 {
		return math.MaxInt64
	}
	return int64(seconds) * 1000
}
